### ISSUE TITLE
Shared memory is not properly cleaned up in some circumstances

### ISSUE THREAD

**Issue by ndrewh** [AFLplusplus/AFLplusplus#2105]
Shared memory before executing afl-showmap:

```
root@e4ce3fb32b83:/out# ipcs -m

------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status      
0x00000000 0          root       600        1048580    3          dest         
0x00000000 1          root       600        8388608    3          dest         
0x00000000 65538      root       600        8388608    0                       
0x00000000 65539      root       600        1048580    0                       
```

Then run afl-showmap... it errors or target crashes, etc. The error doesn't matter. (i assume many different errors can cause this issue -- this is just the one i got).
```
root@e4ce3fb32b83:/out# ./afl-showmap -o tmp -- ./fuzz_target corpus/0034ecacd5427aafc6b97413da2053b36de5059f 
afl-showmap++4.20a by Michal Zalewski
[*] Executing './fuzz_target'...
FS_ERROR_MAP_SIZE

[-] PROGRAM ABORT : AFL_MAP_SIZE is not set and fuzzing target reports that the required size is very large. Solution: Run the fuzzing target stand-alone with the environment variable AFL_DEBUG=1 set and set the value for __afl_final_loc in the AFL_MAP_SIZE environment variable for afl-fuzz.
         Location : report_error_and_exit(), src/afl-forkserver.c:478

```

Memory regions after:
```
root@e4ce3fb32b83:/out# ipcs -m

------ Shared Memory Segments --------
key        shmid      owner      perms      bytes      nattch     status      
0x00000000 0          root       600        1048580    3          dest         
0x00000000 1          root       600        8388608    3          dest         
0x00000000 65538      root       600        8388608    0                       
0x00000000 65539      root       600        1048580    0                       
0x00000000 65540      root       600        8388608    0                       
0x00000000 65541      root       600        1048580    0         
```

The number of memory segments will increase up to the limit (on my machine, it's 4096 in docker), after which shmat will fail, giving:

```
[-]  SYSTEM ERROR : shmget() failed, try running afl-system-config
    Stop location : afl_shm_init(), src/afl-sharedmem.c:284
       OS message : No space left on device
```

tl;dr:
```
for i in $(seq 1 4100); do ./afl-showmap -o tmp -- ./fuzz_target corpus/0034ecacd5427aafc6b97413da2053b36de5059f ; done
```

### Fix?

```
  shmctl(shm->shm_id, IPC_RMID, NULL);
```
will not actually destruct the shared memory until the last process detaches... should we just call this immediately after shmat? That way, if we crash or whatever then the shared memory region will still be destroyed.

**Comment by vanhauser-thc** [comment]
thanks for reporting! fixed in a3125c38f496979be314c93f10adfad9dd4d363e
